gocourse02/main.go

```diff
@@ -0,0 +1,146 @@
+package main
+
+import (
+	"fmt"
+	"math/rand"
+	"os"
+	"time"
+)
+
+/*
+Написати програму “Зоопарк”. Звіри повтікали (більше трьох штук), наглядач повинен їх зібрати.
+Кожна сутність (наглядач, звір, клітка, …) представляється окремою структурою (zookeeper, animal, cage, …).
+Треба використати щонайменше: структури, вказівник, nil, будування, конструктор. Додати тваринам можливість
+розмножуватись. Програма має демонструвати свою роботу через вивід в stdout.
+*/
+
+const animalNumber = 15 // Кількість звірів які повтікали
+
+type ZooKeeper struct {
+	Name           string
+	NumberOfFound  int
+	NumberOfCaught int
+}
+
+func (k *ZooKeeper) Begin(catch Catcher, cage *Cage) {
+	if catch.FindEscapedAnimal() {
+		k.NumberOfFound++
+		if catch.AttemptСatchAnimal(cage) {
+			k.NumberOfCaught++
+			cage.CurrentSeats++
+		}
+	}
+}
+
+type Catcher interface {
+	FindEscapedAnimal() bool       // Спроба знайти звіра
+	AttemptСatchAnimal(*Cage) bool // Сброба зловити звіра в клітку
+}
+
+type Animal struct {
+	ID     int
+	Name   string
+	Gender string
+	Weight int
+	OurNil *int // Вимога задачі використати nil
+	Cage        // Так ми розуміємо чи спіймали звіра
+}
+
+type Cage struct {
+	MaxSeats      int
+	MaxWeight     int
+	CurrentSeats  int
+	CurrentWeight int
+}
+
+// Шукаємо звіра. Результат = Знаходимо або ні
+func (a *Animal) FindEscapedAnimal() bool {
+	return RandBool()
+}
+
+// Намагаемося зловити в клітку знайденого звіра, якщо в клітці є місця  і не перевищує максимально дозволену вагу
+func (a *Animal) AttemptСatchAnimal(cage *Cage) bool {
+	// TODO: перевірити чи  правильні тут розрахунки
+	if cage.MaxWeight >= a.Weight+cage.CurrentWeight && cage.MaxSeats >= cage.CurrentSeats+1 {
+		cage.CurrentWeight += a.Weight
+		a.Cage = *cage
+		return true
+	}
+	return false
+}
+
+// Функція розможноження
+func Reproduction(male, famale *Animal) {
+	fakeData := GetFaker()
+	avgWeight := (male.Weight + famale.Weight) / 2
+	NewAnimals := &Animal{ID: animalNumber + 1, Name: fakeData.Name, Gender: fakeData.Gender, Weight: avgWeight, Cage: Cage{male.MaxSeats, male.MaxWeight, male.CurrentSeats + 1, male.CurrentWeight + avgWeight}}
+
+	_, err := fmt.Fprintf(os.Stdout, "Replay between %s and %s is complete!! Congratulations to the new animal %v \n", male.Name, famale.Name, *NewAnimals)
```

Comment:
Пропоную спростити:
```suggestion
	fmt.Printf("Replay between %s and %s is complete!! Congratulations to the new animal %v \n", male.Name, famale.Name, *NewAnimals)
```

Помилку з [`Printf`](https://pkg.go.dev/fmt#Printf) можна не обробляти. В загальному, ця функція ніколи не повертає помилки. Теоретично це можливо, але ось на практиці дуже рідко таке трапляється.